modules/apps/transfer/types/packet.go

```diff
@@ -204,3 +206,56 @@ func (ftpd FungibleTokenPacketDataV2) GetPacketSender(sourcePortID string) strin
 func (ftpd FungibleTokenPacketDataV2) HasForwarding() bool {
 	return len(ftpd.Forwarding.Hops) > 0
 }
+
+// UnmarshalPacketData attempts to unmarshal the provided packet data bytes into a FungibleTokenPacketDataV2.
+// The version of ics20 should be provided and should be either ics20-1 or ics20-2.
+func UnmarshalPacketData(bz []byte, ics20Version string) (FungibleTokenPacketDataV2, error) {
+	switch ics20Version {
+	case V1:
+		var datav1 FungibleTokenPacketData
+		if err := json.Unmarshal(bz, &datav1); err != nil {
+			return FungibleTokenPacketDataV2{}, errorsmod.Wrapf(ibcerrors.ErrInvalidType, "cannot unmarshal ICS20-V1 transfer packet data: %s", err.Error())
+		}
+
+		return PacketDataV1ToV2(datav1)
+	case V2:
+		var datav2 FungibleTokenPacketDataV2
+		if err := unknownproto.RejectUnknownFieldsStrict(bz, &datav2, unknownproto.DefaultAnyResolver{}); err != nil {
+			return FungibleTokenPacketDataV2{}, errorsmod.Wrapf(ibcerrors.ErrInvalidType, "cannot unmarshal ICS20-V2 transfer packet data: %s", err.Error())
+		}
+
+		if err := proto.Unmarshal(bz, &datav2); err != nil {
+			return FungibleTokenPacketDataV2{}, errorsmod.Wrapf(ibcerrors.ErrInvalidType, "cannot unmarshal ICS20-V2 transfer packet data: %s", err.Error())
+		}
+
+		if err := datav2.ValidateBasic(); err != nil {
+			return FungibleTokenPacketDataV2{}, err
+		}
+
+		return datav2, nil
+	default:
+		return FungibleTokenPacketDataV2{}, errorsmod.Wrap(ErrInvalidVersion, ics20Version)
+	}
+}
+
+// PacketDataV1ToV2 converts a v1 packet data to a v2 packet data. The packet data is validated
+// before conversion.
+func PacketDataV1ToV2(packetData FungibleTokenPacketData) (FungibleTokenPacketDataV2, error) {
```

Comment:
The only difference is that I also made this public. Partially for the test, but I figured it is not actually a bad function to have available? Can undo and delete the tests or do this somehow differently if people want that :)

Reply:
I see the argument for making it public too. As a middleware dev I want to support v2 packets with some update, but also need to support v1. I guess its the same reasoning as us having the function to convert to v2 and maintain one streamlined code path handling v2 data type 🤔